File: C++/Easy/110. 平衡二叉树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(!root) return true;
        if(abs(getTreeHeight(root->left)-getTreeHeight(root->right))>1) return false;
        return isBalanced(root->left)&&isBalanced(root->right);
    }
    int getTreeHeight(TreeNode* root) {
        if(mp.find(root)!=mp.end()) return mp[root];
        if(!root) return 0;
        int h=1;
        mp[root]=h+max(getTreeHeight(root->left),getTreeHeight(root->right));
        return mp[root];
    }
private:
    unordered_map<TreeNode*,int> mp;
};

File: python/Easy/110. 平衡二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def getDepth(root: TreeNode) -> int:
            if root == None:
                return 0
            else:
                return max(getDepth(root.left),getDepth(root.right))+1
        if root == None:
            return True
        leftDepth = getDepth(root.left)
        rightDepth = getDepth(root.right)
        if abs(leftDepth-rightDepth) > 1:
            return False
        else:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
